extract_sets/training_sets.py

```diff
@@ -0,0 +1,398 @@
+# -*- coding: utf-8 -*-
+"""Generate TrainingSets Sets"""
+
+import abc
+import os
+from pathlib import (
+    Path
+)
+from functools import reduce
+import sys
+
+import exifread
+import lxml.etree as etree
+import numpy as np
+
+from cv2 import cv2
+from PIL import (
+    Image
+)
+from shapely.geometry import (
+    Polygon
+)
+
+
+XML_NS = {
+    'alto3': 'http://www.loc.gov/standards/alto/ns-v3#',
+    'alto4': 'http://www.loc.gov/standards/alto/ns-v4#',
+    'page2013': 'http://schema.primaresearch.org/PAGE/gts/pagecontent/2013-07-15',
+    'page2019': 'http://schema.primaresearch.org/PAGE/gts/pagecontent/2019-07-15'}
+
+# default values
+DEFAULT_MIN_CHARS = 1
+DEFAULT_OUTDIR_PREFIX = 'training_data_'
+DEFAULT_USE_SUMMARY = False
+DEFAULT_USE_REORDER = False
+DEFAULT_DPI = 300
+SUMMARY_SUFFIX = '_summary.gt.txt'
+
+# clear unwanted marks for single wordlike tokens
+CLEAR_MARKS = [
+    u'\u200f',  # 'RIGHT-TO-LEFT-MARK'
+    u'\u200e',  # 'LEFT-TO-RIGHT-MARK'
+    u'\ufeff'   # 'ZERO WIDTH NO-BREAK SPACE', the char formerly known as 'BOM'
+]
+
+
+class TextLine(abc.ABC):
+    """
+    TextLine from structured OCR-Data
+    """
+
+    def __init__(self, element, namespace):
+        self.element = element
+        self.namespace = namespace
+        self.element_id = None
+        self.valid = True
+        self.text_words = []
+        self.reorder = None
+
+    @abc.abstractmethod
+    def set_id(self):
+        """Determine identifier"""
+
+    @abc.abstractmethod
+    def set_text(self):
+        """Determine list of word tokens"""
+
+    def get_shape(self, _):
+        """
+        Return TextLine shape
+        Optional(PAGE): Box filled with median color value
+        or greyscale tone to fit rectangular shape
+        """
+
+    def get_textline_content(self) -> str:
+        """
+        Set TextLine contents from it's included word tokens
+        reorder order of tokens if required
+        """
+
+        aggregat = ' '.join(self.text_words)
+        if self.reorder:
+            return reduce(lambda c, p: p + ' ' + c, self.text_words)
+        return aggregat
+
+    def __repr__(self):
+        return '{}[{}]:{}'.format(self.__class__.__name__, self.element_id, self.get_textline_content())
+
+
+class ALTOLine(TextLine):
+    """Extract TextLine Information from ALTO Data"""
+
+    def __init__(self, element, namespace):
+        super().__init__(element, namespace)
+        self.set_id()
+        self.set_text()
+        if self.valid:
+            self.shape = self.get_shape(self.element)
+
+    def set_id(self):
+        self.element_id = self.element.attrib['ID']
+
+    def set_text(self):
+        strings = self.element.findall(f'{self.namespace}:String', XML_NS)
+        self.text_words = [e.attrib['CONTENT'] for e in strings]
```

Comment:
What about the final hyphenation character if it exists?

```suggestion
        strings = self.element.findall(f'{self.namespace}:String', XML_NS)
        strings.extend(self.element.findall(f'{self.namespace}:HYP', XML_NS)
        self.text_words = [e.attrib['CONTENT'] for e in strings]
```

Reply:
You're right, I didn't take this into account yet.